File: java/base/src/main/java/vsdk/toolkit/galerkin/processing/visitors/MaximumRadianceVisitor.java
package vsdk.toolkit.galerkin.processing.visitors;

import vsdk.toolkit.common.color.ColorRgb;
import vsdk.toolkit.galerkin.GalerkinElement;
import vsdk.toolkit.galerkin.GalerkinIterationMethod;
import vsdk.toolkit.galerkin.GalerkinState;

public final class MaximumRadianceVisitor implements ClusterLeafVisitor {
    private ColorRgb accumulatedRadiance;

    public MaximumRadianceVisitor() {
        accumulatedRadiance = new ColorRgb();
        accumulatedRadiance.clear();
    }

    public ColorRgb getAccumulatedRadiance() {
        return accumulatedRadiance;
    }

    @Override
    public void visit(
        GalerkinElement galerkinElement,
        GalerkinState galerkinState)
    {
        ColorRgb rad;
        if ( galerkinState.galerkinIterationMethod == GalerkinIterationMethod.GAUSS_SEIDEL ||
             galerkinState.galerkinIterationMethod == GalerkinIterationMethod.JACOBI ) {
            rad = galerkinElement.radiance[0];
        }
        else {
            rad = galerkinElement.unShotRadiance[0];
        }
        accumulatedRadiance.maximum(accumulatedRadiance, rad);
    }
}

File: cpp/base/src/main/vsdk/toolkit/galerkin/processing/visitors/MaximumRadianceVisitor.cpp
#include "vsdk/toolkit/galerkin/processing/visitors/MaximumRadianceVisitor.h"
MaximumRadianceVisitor::MaximumRadianceVisitor() {
    accumulatedRadiance.clear();
}

MaximumRadianceVisitor::~MaximumRadianceVisitor() {
}

ColorRgbMutable
MaximumRadianceVisitor::getAccumulatedRadiance() const {
    return accumulatedRadiance;
}

void
MaximumRadianceVisitor::visit(
    GalerkinElement *galerkinElement,
    const GalerkinState *galerkinState)
{
    ColorRgbMutable rad(0.0, 0.0, 0.0);
    if ( galerkinState->galerkinIterationMethod == GalerkinIterationMethod::GAUSS_SEIDEL ||
         galerkinState->galerkinIterationMethod == GalerkinIterationMethod::JACOBI ) {
        rad = galerkinElement->radiance[0];
    } else {
        rad = galerkinElement->unShotRadiance[0];
    }
    accumulatedRadiance.maximum(accumulatedRadiance, rad);
}
